pubsub/acks_test.go

```diff
@@ -0,0 +1,229 @@
+	ctx := context.Background()
+	var sentAcks []driver.AckID
+	f := func(ctx context.Context, ackIDs []driver.AckID) error {
+		sentAcks = ackIDs
+		return nil
+	}
+	id := rand.Int()
+	m := &driver.Message{AckID: id}
+	ds := &ackingDriverSub{
+		q:        []*driver.Message{m},
+		sendAcks: f,
+	}
+	sub := pubsub.NewSubscription(ctx, ds, pubsub.SubscriptionOptions{})
+	m2, err := sub.Receive(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := m2.Ack(ctx); err != nil {
+		t.Fatal(err)
+	}
+	if len(sentAcks) != 1 {
+		t.Fatalf("len(sentAcks) = %d, want exactly 1", len(sentAcks))
+	}
+	if sentAcks[0] != id {
+		t.Errorf("sentAcks[0] = %d, want %d", sentAcks[0], id)
+	}
+	if err := sub.Close(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMultipleAcksCanGoIntoASingleBatch(t *testing.T) {
+	ctx := context.Background()
+	sentAcks := make(map[driver.AckID]int)
+	f := func(ctx context.Context, ackIDs []driver.AckID) error {
+		for _, id := range ackIDs {
+			sentAcks[id]++
+		}
+		return nil
+	}
+	ids := []int{rand.Int(), rand.Int()}
+	ds := &ackingDriverSub{
+		q:        []*driver.Message{{AckID: ids[0]}, {AckID: ids[1]}},
+		sendAcks: f,
+	}
+	sopts := pubsub.SubscriptionOptions{
+		AckBatchSize: 2,
+	}
+	sub := pubsub.NewSubscription(ctx, ds, sopts)
+
+	// Receive and ack the messages concurrently.
+	var wg sync.WaitGroup
+	recv := func() {
+		defer wg.Done()
+		mr, err := sub.Receive(ctx)
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		if err := mr.Ack(ctx); err != nil {
+			t.Error(err)
+			return
+		}
+	}
+	wg.Add(2)
+	go recv()
+	go recv()
+	wg.Wait()
+
+	if len(sentAcks) != 2 {
+		t.Errorf("len(sentAcks) = %d, want exactly 2", len(sentAcks))
+	}
+	for _, id := range ids {
+		if sentAcks[id] != 1 {
+			t.Errorf("sentAcks[%v] = %d, want 1", id, sentAcks[id])
+		}
+	}
+	if err := sub.Close(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestTooManyAcksForASingleBatchGoIntoMultipleBatches(t *testing.T) {
+	ctx := context.Background()
+	var sentAckBatches [][]driver.AckID
+	f := func(ctx context.Context, ackIDs []driver.AckID) error {
+		sentAckBatches = append(sentAckBatches, ackIDs)
+		return nil
+	}
+	ids := []int{rand.Int(), rand.Int()}
+	ds := &ackingDriverSub{
+		q:        []*driver.Message{{AckID: ids[0]}, {AckID: ids[1]}},
+		sendAcks: f,
+	}
+	sopts := pubsub.SubscriptionOptions{
+		AckBatchSize: 1,
+	}
+	sub := pubsub.NewSubscription(ctx, ds, sopts)
+
+	// Receive and ack the messages concurrently.
+	var wg sync.WaitGroup
+	recv := func() {
+		mr, err := sub.Receive(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if err := mr.Ack(ctx); err != nil {
+			t.Fatal(err)
+		}
+		wg.Done()
+	}
+	wg.Add(2)
+	go recv()
+	go recv()
+	wg.Wait()
+
+	want := [][]driver.AckID{
+		{ids[0]},
+		{ids[1]},
+	}
+	if !reflect.DeepEqual(sentAckBatches, want) {
+		t.Errorf("got %v, want %v", sentAckBatches, want)
+	}
+
+	if err := sub.Close(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMsgAckReturnsErrorFromSendAcks(t *testing.T) {
```

Comment:
Note: it is interesting to have these tests of the concrete type, but I suspect that at least some of them will end up being conformance tests as well, which may test both concrete type + driver together. At that point they could be removed from here.

Reply:
I think these are worthwhile tests even if there are conformance tests that are similar, but if they turn out to be redundant or burdensome then by all means we should delete them.